TRUNK-5067: Added tests to MessageServiceImpl

pidsonn opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'TRUNK-5111: Replace use of deprecated isVoided' -->
<!--- 'TRUNK-JiraIssueNumber: JiraIssueTitle' -->
## Description of what I changed
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->

## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first! -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it! -->
<!--- Just add the issue number at the end: -->
see https://issues.openmrs.org/browse/TRUNK-5067

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [ x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

pidsonn committed:
TRUNK-5067: Added tests to MessageServiceImpl

pidsonn commented:
https://openmrs.atlassian.net/browse/TRUNK-5067

mherman22 commented on api/src/test/java/org/openmrs/notification/MessageServiceTest.java:
```diff
@@ -1,116 +1,173 @@
-/**
```
did you intentionally remove the license header?

pidsonn replied:
Do you mean this one ->                                                                                                                                                           
 /**
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 *
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */

mherman22 replied:
Yes

pidsonn replied:
No,   i didn't intend it.  May be the file never had that header, but it's ok. Let me fix it.                                                                                  Thanks!

Sanskritii09 commented on api/src/test/java/org/openmrs/notification/MessageServiceTest.java:
```diff
@@ -1,116 +1,173 @@
+import org.openmrs.notification.MessageSender;
 
-/**
- * Unit tests for the MessageService.
- */
 public class MessageServiceTest extends BaseContextSensitiveTest {
 	
-	private static final String NO_SMTP_SERVER_ERROR = "Could not connect to SMTP host:";
-	
-	MessageService ms = null;
+	private MessageServiceImpl messageService;
 	
-	/**
-	 * Run this before each unit test in this class. The "@Before" method in
-	 * {@link BaseContextSensitiveTest} is run right before this method.
-	 * 
```
✅ Good improvements: Added unit tests, proper test setup, and integration with OpenMRS test data.
⚠️ Needs minor fixes: Add missing assertions, avoid redundant initialization, and refactor hardcoded strings.

Sanskritii09 commented on api/src/test/java/org/openmrs/notification/MessageServiceTest.java:
```diff
@@ -1,116 +1,173 @@
- * graphic logo is a trademark of OpenMRS Inc.
- */
 package org.openmrs.notification;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.fail;
 
+import java.util.Map;
+
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.openmrs.api.context.Context;
 import org.openmrs.test.jupiter.BaseContextSensitiveTest;
+import org.openmrs.notification.MessageServiceImpl;
+import org.openmrs.notification.MessageException;
+import org.openmrs.notification.Message;
+import org.openmrs.notification.Template;
+import org.openmrs.notification.MessagePreparator;
+import org.openmrs.notification.MessageSender;
 
-/**
- * Unit tests for the MessageService.
- */
 public class MessageServiceTest extends BaseContextSensitiveTest {
 	
-	private static final String NO_SMTP_SERVER_ERROR = "Could not connect to SMTP host:";
-	
-	MessageService ms = null;
+	private MessageServiceImpl messageService;
 	
-	/**
-	 * Run this before each unit test in this class. The "@Before" method in
```
✅ Good improvements: Added unit tests, proper test setup, and integration with OpenMRS test data.
⚠️ Needs minor fixes: Add missing assertions, avoid redundant initialization, and refactor hardcoded strings.

pidsonn replied:
Thank u 4 ur time and advice.                                                                                                                                                                                                I'm busy polishing these codes but i have one challenge >> The openmrs initialisation wizard never completed successfully. I have tried it many times but still fail. THEN, about these tests. Yes i made some fixes here and there but the problem is that these tests keep passing and failing on different trials i don't know why??  could it be due to that initialisation wizard issue i mentioned earlier or something else. Now, can i push these changes as they are? or i wait until they all pass at all times, because i'm still investigating.                                                                                        Thanx.

pidsonn committed:
TRUNK-5067: Updated tests and added the license header

pidsonn commented:
I've made the requested changes, including test updates and adding the license header. Let me know if anything else is needed.

Thanks!

pidsonn reviewed (commented):
I've made the requested changes, including test updates and adding the license header. Let me know if anything else is needed.
Thanks!

pidsonn reviewed (commented):
Hi @ibacher or @dkayiwa, I’ve updated the branch with the latest from master. Kindly requesting workflow approval and a review when convenient. Thank you for your time!

pidsonn committed:
Add applicationContext-test.xml and update MessageServiceTest for test context

pidsonn commented on api/src/main/java/org/openmrs/api/context/Context.java:
```diff
@@ -270,7 +269,7 @@ public static UserContext getUserContext() {
 	 * @return the current ServiceContext
 	 */
-	static ServiceContext getServiceContext() {
+	public static ServiceContext getServiceContext() {
```
fixed!

pidsonn commented on api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java:
```diff
@@ -814,5 +816,8 @@ public void changePasswordUsingActivationKey(String activationKey, String newPas
 	public String getLastLoginTime(User user) {
 		return dao.getLastLoginTime(user);
-	}
+    }
```
Resolved!

pidsonn commented on api/src/main/java/org/openmrs/notification/impl/MessageServiceImpl.java:
```diff
@@ -89,6 +102,7 @@ public MessageSender getMessageSender() {
 	 */
 	@Override
 	public void sendMessage(Message message) throws MessageException {
+		mailMessageSender.send(message);
```
Sorted!

pidsonn committed:
TRUNK-5067: Removed unnecessary @ContextConfiguration; BaseContextSensitiveTest already provides required context. PR now focuses solely on testing MessageServiceImpl as it exists.

pidsonn commented:
✅ Update: Thanks, @ibacher — I’ve refactored this PR to focus solely on testing MessageServiceImpl as it exists.

    . Removed the unnecessary @ContextConfiguration; BaseContextSensitiveTest already provides the required context.

    . Rebased on master and force-pushed.

pidsonn committed:
TRUNK-5067:Add Javadoc comments to MessageServiceTest methods

pidsonn committed:
Merge branch 'master' into TRUNK-5067

pidsonn commented:
Hi @ibacher,
 I’ve confirmed that the failing CI checks (Maven build) are not due to my changes. Since this is beyond my contribution scope, I’d appreciate your guidance on how to proceed.
 Thanks!

pidsonn committed:
Merge branch 'master' into TRUNK-5067

pidsonn committed:
Merge branch 'master' into TRUNK-5067

pidsonn committed:
Merge branch 'master' into TRUNK-5067

pidsonn committed:
Merge branch 'master' into TRUNK-5067

ibacher committed:
Discard changes to api/src/main/java/org/openmrs/api/context/Context.java

ibacher committed:
Discard changes to api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java

pidsonn committed:
Fix build failures.

pidsonn committed:
Merge branch 'master' into TRUNK-5067

ibacher commented on api/src/main/java/org/openmrs/notification/impl/MessageServiceImpl.java:
```diff
@@ -24,37 +24,45 @@
+    private MessageService self;
+
+
+
```
Why are there so many empty lines here?

ibacher commented on api/src/main/java/org/openmrs/notification/impl/MessageServiceImpl.java:
```diff
@@ -64,12 +72,13 @@ public MessageServiceImpl() {
 	}
-	
+
+
```
Why are we adding another blank line here?

ibacher commented on api/src/test/java/org/openmrs/notification/MessageServiceTest.java:
```diff
@@ -7,110 +7,294 @@
  * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
  * graphic logo is a trademark of OpenMRS Inc.
  */
+
```
Please remove

ibacher commented on api/src/test/java/org/openmrs/notification/MessageServiceTest.java:
```diff
@@ -7,110 +7,294 @@
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.times;
```
All of the static imports should be grouped together

ibacher commented on api/src/test/java/org/openmrs/notification/MessageServiceTest.java:
```diff
@@ -7,110 +7,294 @@
 
-/**
- * Unit tests for the MessageService.
- */
```
Why is this being removed?

ibacher commented on api/src/test/java/org/openmrs/notification/MessageServiceTest.java:
```diff
@@ -7,110 +7,294 @@
+    selfField.set(messageService, messageService);
+ } catch (NoSuchFieldException | IllegalAccessException e) {
+     throw new RuntimeException("Failed to set 'self' field via reflection", e);
+ }
```
Also, we should just autowire the thing instead of using reflection here. That way the test actually, you know, tests the code.

ibacher commented on api/src/test/java/org/openmrs/notification/MessageServiceTest.java:
```diff
@@ -7,110 +7,294 @@
-		
-		assertEquals(attachment, msg4.getAttachment());
-		assertEquals(attachmentContentType, msg4.getAttachmentContentType());
-		assertEquals(attachmentFileName, msg4.getAttachmentFileName());
```
I'm fine with re-writing the tests into a single-code-path-per test. That makes sense. However, you've now actually removed coverage for everything except one function. We should also have a test for the other variations of `createMessage()`, e.g., `createMessage_shouldCreateMessageWithJustSubjectAndMessage()`, etc.

ibacher commented on api/src/test/java/org/openmrs/notification/MessageServiceTest.java:
```diff
@@ -7,110 +7,294 @@
+    /**
+     * Tests that the message sender successfully sends a basic message.
+     *
+     * @throws Exception if sending fails
```
This is actually less useful than the Javadoc it replaces.

ibacher commented on api/src/test/java/org/openmrs/notification/MessageServiceTest.java:
```diff
@@ -7,110 +7,294 @@
+     * Ensures that the message service can prepare a message from a given template.
+     *
+     * @throws Exception if preparation fails
+     */
```
Formatting

ibacher commented on api/src/test/java/org/openmrs/notification/MessageServiceTest.java:
```diff
@@ -7,110 +7,294 @@
+    Template template = new Template();
+
+    when(messagePreparator.prepare(any(Template.class)))
+         .thenAnswer(invocation -> new Message());
```
```suggestion
         .thenAnswer(Message::new);
```

ibacher commented on api/src/test/java/org/openmrs/notification/MessageServiceTest.java:
```diff
@@ -7,110 +7,294 @@
+    expectedMessage.setRecipients("user@example.com");
+
+    // ✅ Tell the mock what to return
+    when(messagePreparator.prepare(template)).thenReturn(expectedMessage);
```
Why do we retain the mock above if we shifted to this? Also, why have we dropped the arrange, act, assert structure here?

ibacher commented on api/src/test/resources/TestingApplicationContext.xml:
```diff
@@ -65,4 +65,11 @@
 		<property name="transactionAttributeSource" ref="transactionAttributeSource"/>
 	</bean>
+	<!-- Mock MailMessageSender to fix test error -->
+	<bean id="mailMessageSender" class="org.mockito.Mockito" factory-method="mock">
```
Also please pay attention to how files are already formatted when adding things to them.

ibacher reviewed (changes requested):
Thanks @pidsonn! This still needs work, but it's getting there.

pidsonn committed:
TRUNK-5067: Update MessageServiceImpl, tests, TestingApplicationContext.xml and delete applicationContext-test.xml

pidsonn committed:
Merge branch 'master' into TRUNK-5067
